bitswap/network/router.go

```diff
@@ -0,0 +1,136 @@
+	}
+
+	if http == nil {
+		return bitswap
+	}
+
+	if bitswap == nil {
+		return http
+	}
+
+	return &router{
+		Peerstore: pstore,
+		Bitswap:   bitswap,
+		HTTP:      http,
+	}
+}
+
+func (rt *router) Start(receivers ...Receiver) {
+	rt.Bitswap.Start(receivers...)
+	rt.HTTP.Start(receivers...)
+}
+
+func (rt *router) Stop() {
+	rt.Bitswap.Stop()
+	rt.HTTP.Stop()
+}
+
+// Should be the same for both
+func (rt *router) Self() peer.ID {
+	return rt.Bitswap.Self()
+}
+
+func (rt *router) Ping(ctx context.Context, p peer.ID) ping.Result {
+	pi := rt.Peerstore.PeerInfo(p)
+	htaddrs, _ := SplitHTTPAddrs(pi)
+	if len(htaddrs.Addrs) > 0 {
+		return rt.HTTP.Ping(ctx, p)
+	}
+	return rt.Bitswap.Ping(ctx, p)
+}
+
+func (rt *router) Latency(p peer.ID) time.Duration {
+	pi := rt.Peerstore.PeerInfo(p)
+	htaddrs, _ := SplitHTTPAddrs(pi)
+	if len(htaddrs.Addrs) > 0 {
+		return rt.HTTP.Latency(p)
+	}
+	return rt.Bitswap.Latency(p)
+}
+
+func (rt *router) SendMessage(ctx context.Context, p peer.ID, msg bsmsg.BitSwapMessage) error {
+	// SendMessage is only used by bitswap server so we send a bitswap
+	// message.
+	return rt.Bitswap.SendMessage(ctx, p, msg)
+}
+
+// Connect attempts to connect to a peer. It prioritizes HTTP connections over
+// bitswap.
+func (rt *router) Connect(ctx context.Context, p peer.AddrInfo) error {
+	htaddrs, bsaddrs := SplitHTTPAddrs(p)
+	err := rt.HTTP.Connect(ctx, htaddrs)
+	if err != nil {
+		return rt.Bitswap.Connect(ctx, bsaddrs)
+	}
+	return nil
+}
+
+func (rt *router) DisconnectFrom(ctx context.Context, p peer.ID) error {
+	return multierr.Combine(
+		rt.HTTP.DisconnectFrom(ctx, p),
+		rt.Bitswap.DisconnectFrom(ctx, p),
+	)
+}
+
+func (rt *router) Stats() Stats {
+	htstats := rt.HTTP.Stats()
+	bsstats := rt.Bitswap.Stats()
+	return Stats{
+		MessagesRecvd: htstats.MessagesRecvd + bsstats.MessagesRecvd,
+		MessagesSent:  htstats.MessagesSent + bsstats.MessagesSent,
+	}
+}
+
+// NewMessageSender returns a MessageSender using the HTTP network when HTTP
+// addresses are known, and bitswap otherwise.
+func (rt *router) NewMessageSender(ctx context.Context, p peer.ID, opts *MessageSenderOpts) (MessageSender, error) {
+	// IF we did not manage to connect to any HTTP address beforehand, we
+	// should not have them in the peerstore.
+	pi := rt.Peerstore.PeerInfo(p)
+	htaddrs, _ := SplitHTTPAddrs(pi)
+	if len(htaddrs.Addrs) > 0 {
+		return rt.HTTP.NewMessageSender(ctx, p, opts)
+	}
+	return rt.Bitswap.NewMessageSender(ctx, p, opts)
+}
+
+func (rt *router) TagPeer(p peer.ID, tag string, w int) {
+	rt.HTTP.TagPeer(p, tag, w)
+	rt.Bitswap.TagPeer(p, tag, w)
+}
+
+func (rt *router) UntagPeer(p peer.ID, tag string) {
+	rt.HTTP.UntagPeer(p, tag)
+	rt.Bitswap.UntagPeer(p, tag)
+}
+
+func (rt *router) Protect(p peer.ID, tag string) {
+	rt.HTTP.Protect(p, tag)
+	rt.Bitswap.Protect(p, tag)
+}
+func (rt *router) Unprotect(p peer.ID, tag string) bool {
+	return rt.HTTP.Unprotect(p, tag) || rt.Bitswap.Unprotect(p, tag)
+}
```

Comment:
If we implement something akin to `Protect` in HTTP as mentioned above, we might want to check this.

If we prioritize HTTP for peers with known HTTP addresses, we might not want to protect bitswap connections because they should not exist in the first place. Or perhaps they should when the server is using them to provide blocks via Bitswap.

Reply:
New approach is to follow prioritization rules.